systemd: switch to unified cgroup hierarchy by default

flokli opened the pull request:
Other distros, like Fedora 32 have already switched to it.

Users that need to keep using the old cgroup hierarchy (namely because
they want to run docker, and not some of the alternatives supporting
cgroupsv2 can re-enable it by setting `systemd.enableUnifiedCgroupHierarchy = false;`.

Fixes #73800

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

flokli committed:
systemd: switch to unified cgroup hierarchy by default

See https://www.redhat.com/sysadmin/fedora-31-control-group-v2 for
details on why this is desirable, and how it impacts containers.

Users that need to keep using the old cgroup hierarchy can re-enable it
by setting `systemd.unifiedCgroupHierarchy` to `false`.

Well-known candidates not supporting that hierarchy, like docker and
hidepid=… will disable it automatically.

Fixes #73800

grahamc commented on nixos/doc/manual/release-notes/rl-2103.xml:
```diff
@@ -264,6 +264,19 @@
    </listitem>
+   <listitem>
+    <para>
+     NixOS now defaults to the unified cgroup hierarchy (cgroupsv2), which has been made a default
+     in other distributions, like Fedora 32.
+     See the <link xlink:href="https://www.redhat.com/sysadmin/fedora-31-control-group-v2">Fedora Article for 32</link>
+     for details on why this is desirable, and how it impacts containers.
```
I'm not sure we need to appeal to other distros as an "authority".  How about something like:

```suggestion
     NixOS now defaults to the unified cgroup hierarchy (cgroupsv2).
     See the <link xlink:href="https://www.redhat.com/sysadmin/fedora-31-control-group-v2">Fedora Article for 32</link>
     for details on why this is desirable, and how it impacts containers.
```

flokli replied:
Updated, thanks!

grahamc reviewed (commented):
I have this in my config:

```
   boot.kernelParams = (if config.virtualisation.docker.enable then [] else [ "cgroup_no_v1=all" "systemd.unified_cgroup_hierarchy=yes" ]);
```

maybe we should automatically specify this for the docker services?

andir commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -281,9 +281,6 @@ stdenv.mkDerivation {
     "-Dcreate-log-dirs=false"
-    # Upstream uses cgroupsv2 by default. To support docker and other
-    # container managers we still need v1.
-    "-Ddefault-hierarchy=hybrid"
```
Ugh, why not be explicit? Both here and via the module system. If we have this in the module system we can throw proper warnings/errors when docker is enabled.

flokli replied:
Makes sense, done.

flokli reviewed (commented):
> I have this in my config: 
```
boot.kernelParams = (if config.virtualisation.docker.enable then [] else [ "cgroup_no_v1=all" "systemd.unified_cgroup_hierarchy=yes" ]);
```

> maybe we should automatically specify this for the docker services?

Yes, adding `"systemd.unified_cgroup_hierarchy=0` if docker is enabled might make sense, at least as long as docker doesn't support cgroupv2.
We probably should add link to the upstream tracking issues.

It seems it's supported in runc: https://github.com/opencontainers/runc/issues/2315#issuecomment-700423212 and should be supported in docker 20, which was "not officially released yet, and its ETA still unpublished" in September 2020: https://github.com/moby/moby/issues/40360#issuecomment-699867043

SuperSandro2000 commented:
Docker 20.10.0-rc1 was released today which should add cgroupsv2 and apparmor 3.0-beta support. Maybe we can wait until it is released?

flokli commented:
@SuperSandro2000 are you aware of any timelines on 20.* getting released? Would you be up to packaging their rc, to see how it behaves? ;-)

I surely want to add/extend some more tests about this before merging, and see how config options for other container runtimes might need to look like (if it requires special configuration).

ajs124 commented:
This will make user@ services fail when using with the hardened profile. Specifically [this line of the hardened profile](https://github.com/NixOS/nixpkgs/blob/b6e21a7/nixos/modules/profiles/hardened.nix#L20) and [this issue](https://github.com/systemd/systemd/issues/12955#issuecomment-508490893) marked as "not supported".

There are several solutions to this and it definitely should not block this, but I thought it was worth noting.

cc @joachifm (because of the hardened profile).

flokli committed:
nixos/k3s: disable unifiedCgroupHierarchy

This gets automatically disabled by docker if the docker backend is
used, but the bundled containerd also doesn't seem to support cgroupsv2,
so disable it explicitly here, too.

flokli commented:
I also did some more experiments:

 - cri-o works fine with unified cgroups.
 - k3s with docker as container runtime obviously still needs unified cgroups disabled
 - k3s with their bundled containerd doesn't work with cgroupsv2 either

I made sure the unified cgroup hierarchy is disabled in these cases and the tests succeed again.

flokli committed:
nixosTests.systemd: increase accounting coverage

For now, testing IO Accounting is skipped, as it seems to be either
broken, or hard to reproduce in a VM.

flokli commented:
I added some more tests on the accounting parts. I couldn't reliably get IO Accounting on individual units tested, but the test ensures `system.slice` has non-zero IO accounting values :-)

This is ready for review, PTAL.

SuperSandro2000 commented:
> are you aware of any timelines on docker 20.* getting released?

No but I assume it could take a month.

> Would you be up to packaging their rc, to see how it behaves?

I can get it to build but for testing I miss experience.

flokli commented:
> > are you aware of any timelines on docker 20.* getting released?
> 
> No but I assume it could take a month.

Okay. So I propose we keep this PR as-is (disable unified cgroup when you enable docker), and we can remove this once docker has landed and supports it.

> > Would you be up to packaging their rc, to see how it behaves?
> 
> I can get it to build but for testing I miss experience.

`nixosTests.docker` should probably be enough to verify basic functionality, and once that works, people could take a more closer look in the PR :-)

SuperSandro2000 commented:
> nixosTests.docker should probably be enough to verify basic functionality, and once that works, people could take a more closer look in that new PR :-)

KVM does not want to work with nix tests on my machine. Not sure why.

ajs124 commented on nixos/tests/systemd.nix:
```diff
@@ -168,5 +161,25 @@ import ./make-test-python.nix ({ pkgs, ... }: {
+
+    with subtest("systemd per-unit accounting works"):
+        assert "IP traffic received: 84B" in output_ping
+        assert "IP traffic sent: 84B" in output_ping
```
Are you sure this won't break?

flokli replied:
It was constant here at least.

ajs124 reviewed (approved):
LGTM, in general. Some notes, although they're not too important.

flokli commented on nixos/modules/security/hidepid.nix:
```diff
@@ -23,5 +23,9 @@ with lib;
+
+    # Disable cgroupsv2, which doesn't work with hidepid.
+    # https://github.com/NixOS/nixpkgs/pull/104094#issuecomment-729996203
+    systemd.unifiedCgroupHierarchy = false;
```
For now, I just disabled it where it caused breakage, so the test kept working without any config change. I'm happy to change things here, but don't think it needs to block this PR.

flokli commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -76,6 +76,10 @@ in
+    # TODO: disable this once k3s supports cgroupsv2, either by docker
+    # supporting it, or their bundled containerd
+    systemd.unifiedCgroupHierarchy = false;
+
```
This currently doesn't work with k3s, and I'd expect people enabling k3s to want to have it working.

arianvp commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -550,6 +550,14 @@ in
       '';
     };
 
+    systemd.unifiedCgroupHierarchy = mkOption {
```
*bikeshed*: Can we call this `enableUnifiedCgroupHierarchy`  so it's the same naming scheme as the other boolean options in this module like `enableCgroupAccounting` ?

flokli replied:
bikeshed accepted and updated.

flokli committed:
nixosTests.podman: run default backends, don't run runc rootless

The runc backend doesn't work with unified cgroup hierarchy, and it
failing is a known issue.

However, the default backends should work in both rootless and as-root
scenarios, so make sure we test these.

zowoq commented on nixos/tests/podman.nix:
```diff
@@ -53,22 +53,11 @@ import ./make-test-python.nix (
-          podman.succeed(
-              su_cmd(
-                  "podman run --runtime=crun -d --name=sleeping -v /nix/store:/nix/store -v /run/current-system/sw/bin:/bin scratchimg /bin/sleep 10"
+                  "podman run -d --name=sleeping -v /nix/store:/nix/store -v /run/current-system/sw/bin:/bin scratchimg /bin/sleep 10"
```
Can we keep the explicit `--runtime=crun`/ `Run container rootless with crun`  please?

flokli replied:
Can you elaborate on why? `podman` seems to default to it with cgroupsv2, and if we view this test from a user perspective, they just want to `podman run` as non-root, without explicitly configuring a specific runtime.

also cc @adisbladis

adisbladis replied:
I just want to say I agree with @flokli on this one. It's better that we test the users default config.
Also I seem to recall seeing some intent to switch back to `runc` by default at some point once it's gained the required cgroupsv2 support (don't quote me on this one).

zowoq replied:
Can we do this instead?

```diff
diff --git a/nixos/tests/podman.nix b/nixos/tests/podman.nix
index cd8c2b4308c..74581f5c847 100644
--- a/nixos/tests/podman.nix
+++ b/nixos/tests/podman.nix
@@ -55,13 +55,13 @@ import ./make-test-python.nix (
 
       with subtest("Run container rootless with runc"):
           podman.succeed(su_cmd("tar cv --files-from /dev/null | podman import - scratchimg"))
-          podman.succeed(
+          podman.fail(
               su_cmd(
                   "podman run --runtime=runc -d --name=sleeping -v /nix/store:/nix/store -v /run/current-system/sw/bin:/bin scratchimg /bin/sleep 10"
               )
           )
-          podman.succeed(su_cmd("podman ps | grep sleeping"))
-          podman.succeed(su_cmd("podman stop sleeping"))
+          podman.fail(su_cmd("podman ps | grep sleeping"))
+          podman.fail(su_cmd("podman stop sleeping"))
           podman.succeed(su_cmd("podman rm sleeping"))
 
       with subtest("Run container rootless with crun"):

```

flokli replied:
I don't think our test should suddenly fail if podmans runc backend starts working in rootless scenarios again.

I'm fine adding a test for the default backend in as-root scenarios, but made the rootless runc subtest a comment, which we can re-add once it gets implemented.

zowoq replied:
> I don't think our test should suddenly fail if podmans runc backend starts working in rootless scenarios again.
> 
> I'm fine adding a test for the default backend in as-root scenarios, but made the rootless runc subtest a comment, which we can re-add once it gets implemented.

All of these packages have `passthru.tests` so the failure will be noticed when it does start working.

Can you move the default tests after the explict runtime tests please? If it does fail because of a runtime problem having it fail on the explict test would be clearer.

flokli replied:
I can move the test order, sure, but I still don't think the test should suddenly fail if *more* things work. `podman` is still functional if a backend that was known to be broken starts working again. We shouldn't then fail the test.

flokli replied:
I'd expect the maintainers of this package to read the release notes while bumping, and updating/extending the test accordingly once new features are added.

andir commented on nixos/modules/virtualisation/docker.nix:
```diff
@@ -155,6 +155,9 @@ in
       systemd.packages = [ cfg.package ];
 
+      # TODO: remove once docker 20.10 is released
+      systemd.enableUnifiedCgroupHierarchy = false;
```
Would it make sense to make this conditional on the docker version? Users can set the docker package at runtime and defaulting this to the right version automatically might be a nice perk that we can implement rather quickly in NixOS?

flokli replied:
We don't know yet if docker 20.10 really works. I'm happy to take a closer look at this once it is in Nixpkgs. There currently is no docker working with enabled unified cgroup hierarchy, so I'm not much in favor of code for the eventuality of it being there ;-)

flokli commented:
Let's get this in. If there's other applications that don't really work with cgroupsv2, disabling it for them is pretty trivial by setting `systemd.enableUnifiedCgroupHierarchy = false;`.
